src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                        {
+                            predecessor.Next = next;
+                            next.AddPredecessor(predecessor);
+                        }
+
+                        (IOperation condition, bool jumpIfTrue, BasicBlock destination) = predecessor.Conditional;
+                        if (destination == block)
+                        {
+                            predecessor.Conditional = (condition, jumpIfTrue, next);
+                            next.AddPredecessor(predecessor);
+                        }
+                    }
+
+                    next.RemovePredecessor(block);
+                    blocks.RemoveAt(i);
+                    i--;
+                    count--;
+                }
+            }
+        }
+
+        private void VisitStatement(IOperation operation)
+        {
+            IOperation saveCurrentStatement = _currentStatement;
+            _currentStatement = operation;
+            Debug.Assert(_evalStack.Count == 0);
+
+            // PROTOTYPE(dataflow): Ensure that statement's parent is null at this point.
+            AddStatement(Visit(operation, null));
+            Debug.Assert(_evalStack.Count == 0);
+            _currentStatement = saveCurrentStatement;
+        }
+
+        private BasicBlock CurrentBasicBlock
+        {
+            get
+            {
+                if (_currentBasicBlock == null)
+                {
+                    AppendNewBlock(new BasicBlock(BasicBlockKind.Block));
+                }
+
+                return _currentBasicBlock;
+            }
+        }
+
+        private void AddStatement(IOperation statement)
+        {
+            if (statement == null)
+            {
+                return;
+            }
+
+            // PROTOTYPE(dataflow): Assert that statement's parent is null at this point.
+            CurrentBasicBlock.AddStatement(statement);
+        }
+
+        private void AppendNewBlock(BasicBlock block)
+        {
+            Debug.Assert(block != null);
+            BasicBlock prevBlock = _blocks.Last();
+
+            if (prevBlock.Next == null)
+            {
+                LinkBlocks(prevBlock, block);
+            }
+
+            _blocks.Add(block);
+            _currentBasicBlock = block;
+        }
+
+        private static void LinkBlocks(BasicBlock prevBlock, BasicBlock nextBlock)
+        {
+            Debug.Assert(prevBlock.Next == null);
+            prevBlock.Next = nextBlock;
+            nextBlock.AddPredecessor(prevBlock);
+        }
+
+        public override IOperation VisitBlock(IBlockOperation operation, object argument)
+        {
+            foreach (var statement in operation.Operations)
+            {
+                VisitStatement(statement);
+            }
+
+            return null;
+        }
+
+        public override IOperation VisitConditional(IConditionalOperation operation, object argument)
+        {
+            if (operation == _currentStatement)
```

Comment:
It seems this is the main purpose of `_currentStatement`, which is to identify if the conditional operation was a statement or merely an expression. Isn't there a more direct way to test that on the operation node, maybe by checking the `Kind` on its parent?

Reply:
> It seems this is the main purpose of _currentStatement, which is to identify if the conditional operation was a statement or merely an expression. Isn't there a more direct way to test that on the operation node, maybe by checking the Kind on its parent?

IMHO this is the simplest reliable way because checking the kind of a parent simply is not sufficient when parent is not a IBlockOperation, for example a for loop, etc.

---
In reply to: [162159877](https://github.com/dotnet/roslyn/pull/24263#discussion_r162159877) [](ancestors = 162159877)